azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+   * @return true if the status is different from cached value, false otherwise
+   */
+  private boolean isUpdatedPodStatusDistinct(AzPodStatusMetadata event) {
+    AzPodStatus currentStatus = AzPodStatus.AZ_POD_UNSET;
+    if (podStatusMap.containsKey(event.getPodName())) {
+      currentStatus = podStatusMap.get(event.getPodName()).getAzPodStatus();
+    }
+    boolean shouldSkip = currentStatus == event.getAzPodStatus();
+    if (shouldSkip) {
+      logger.info(format("Event pod status is same as current %s, for pod %s."
+          +" Any updates will be skipped.", currentStatus, event.getPodName()));
+    }
+    return !shouldSkip;
+  }
+
+  // Update the cache with the given event.
+  private void updatePodStatus(AzPodStatusMetadata event) {
+    podStatusMap.put(event.getPodName(), event);
+    logger.debug(format("Updated status to %s, for pod %s", event.getAzPodStatus(), event.getPodName()));
+  }
+
+  /**
+   * Apply the boolean function {@code expectedStatusMatcher} to execution-id and if it returns true
+   * then finalize the status of the flow in db.
+   * This can be used for finalizing the flow based on whether the current Flow Status has a
+   * specific value or is present within a given set of values.
+   *
+   * <p> Note:
+   * Unfortunately many (if not all) of the Flow status updates in Azkaban don't fully enforce
+   * the Flow lifecycle state-machine and {@code ExecutionControllerUtils.finalizeFlow} used
+   * within this method is no exception. It will simply finalize the flow (as failed) even if the
+   * flow status is in a finalized state in the Db. <br>
+   * This indirectly impacts this listener implementation as occasionally more than one thread
+   * could try to update the state of the same flow in db. Although it's not any worse than how
+   * the rest of Azkaban already behaves, we should fix the behavior at least during the
+   * finalization of flows. One way of achieving this is to add a utility method to atomically
+   * test-and-finalize a flow from a non-final to a failed state.
+   *
+   * @implNote Flow status check and update is not atomic, details above.
+   * @param event
+   * @param expectedStatusMatcher
+   * @return
+   */
+  private azkaban.executor.Status compareAndFinalizeFlowStatus(AzPodStatusMetadata event,
+      Function<Status, Boolean> expectedStatusMatcher) {
+    requireNonNull(expectedStatusMatcher, "flow matcher must not be null");
+
+    int executionId = Integer.parseInt(event.getFlowPodMetadata().get().getExecutionId());
+    ExecutableFlow executableFlow = null;
+    try {
+      executableFlow = executorLoader.fetchExecutableFlow(executionId);
+    } catch (ExecutorManagerException e) {
+      String message = format("Exception while fetching executable flow for pod %s",
+          event.getPodName());
+      logger.error(message, e);
+      throw new AzkabanWatchException(message, e);
+    }
+    checkState(executableFlow != null, "executable flow must not be null");
```

Comment:
Shall we log it silently?

Reply:
I do see the value in logging this silently, however for a valid flow container we `fetchExecutableFlow` should not return null.
I have refactored parts of this code for the next PR, which should address any concerns about this. We can revisit this as part of that PR.